src/IO/AsynchronousReader.h

```diff
@@ -47,29 +50,31 @@ class IAsynchronousReader : private boost::noncopyable
         FileDescriptorPtr descriptor;
         size_t offset = 0;
         size_t size = 0;
+        /// If descriptor is a RemoteFSFileDescriptor containing a CachedInMemoryReadBufferFromFile
+        /// then `buf` can be nullptr, and PageCacheCell buffer will be used instead (to avoid
+        /// copying data out of userspace page cache).
         char * buf = nullptr;
         Priority priority;
         size_t ignore = 0;
     };
 
     struct Result
     {
-        /// The read data is at [buf + offset, buf + size), where `buf` is from Request struct.
+        /// The read data is at [buf + offset, buf + size).
         /// (Notice that `offset` is included in `size`.)
-
-        /// size
+        /// offset is equal to request.ignore.
+        /// buf is either the buf from Request or inside page_cache_cell's buffer,
+        /// or nullptr if size = 0.
         /// Less than requested amount of data can be returned.
         /// If size is zero - the file has ended.
         /// (for example, EINTR must be handled by implementation automatically)
+        char * buf = nullptr;
         size_t size = 0;
-
-        /// offset
-        /// Optional. Useful when implementation needs to do ignore().
         size_t offset = 0;
 
-        std::unique_ptr<Stopwatch> execution_watch = {};
+        PageCacheCellPtr page_cache_cell = {};
```

Comment:
As `page_cache_cell` does not seem to be used anyhow in `AsynchronousBoundedReadBuffer` then we can remove it from here as well